server: TestAdminAPIDatabaseDetails failed

server.TestAdminAPIDatabaseDetails [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10660514?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10660514?buildTab=artifacts#/) on release-22.2 @ [e36a88e4bd22dae1108e1276d6a4cdc680bc6a1b](https://github.com/cockroachdb/cockroach/commits/e36a88e4bd22dae1108e1276d6a4cdc680bc6a1b):


```
=== RUN   TestAdminAPIDatabaseDetails
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/5b2c9b3a394428c7572d34050aad8975/logTestAdminAPIDatabaseDetails1011264950
    test_log_scope.go:79: use -show-logs to present logs inline
*
* INFO: Running test with the default test tenant. If you are only seeing a test case failure when this message appears, there may be a problem with your test case running within tenants.
*
    admin_cluster_test.go:75: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/server_test/pkg/server/admin_cluster_test.go:75
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:69
        	            				github.com/cockroachdb/cockroach/pkg/util/retry/retry.go:213
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:75
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:57
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:40
        	            				github.com/cockroachdb/cockroach/pkg/server_test/pkg/server/admin_cluster_test.go:59
        	Error:      	Not equal: 
        	            	expected: 0x0
        	            	actual  : 0x9b1
        	Test:       	TestAdminAPIDatabaseDetails
        	Messages:   	ApproximateDiskBytes
    admin_cluster_test.go:79: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/5b2c9b3a394428c7572d34050aad8975/logTestAdminAPIDatabaseDetails1011264950
--- FAIL: TestAdminAPIDatabaseDetails (23.48s)
```
<p>Parameters: <code>TAGS=bazel,gss,race</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #103394 server: TestAdminAPIDatabaseDetails failed [A-cluster-observability C-test-failure O-robot T-cluster-observability branch-release-22.1]
</p>
</details>
/cc @cockroachdb/obs-inf-prs @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAdminAPIDatabaseDetails.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-29085

--- a/pkg/server/admin_cluster_test.go
+++ b/pkg/server/admin_cluster_test.go
@@ -16,6 +16,7 @@ import (
 	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/liveness/livenesspb"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/server/serverpb"
@@ -33,7 +34,9 @@ import (
 func TestAdminAPIDatabaseDetails(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
-	tc := testcluster.StartTestCluster(t, 3, base.TestClusterArgs{})
+
+	const numServers = 3
+	tc := testcluster.StartTestCluster(t, numServers, base.TestClusterArgs{})
 	defer tc.Stopper().Stop(context.Background())
 
 	db := tc.ServerConn(0)
@@ -49,13 +52,22 @@ func TestAdminAPIDatabaseDetails(t *testing.T) {
 		require.NoError(t, err)
 	}
 
+	// Flush all stores here so that we can read the ApproximateDiskBytes field without waiting for a flush.
+	for i := 0; i < numServers; i++ {
+		s := tc.Server(i)
+		err = s.GetStores().(*kvserver.Stores).VisitStores(func(store *kvserver.Store) error {
+			return store.TODOEngine().Flush()
+		})
+		require.NoError(t, err)
+	}
+
 	s := tc.Server(0)
 
 	var resp serverpb.DatabaseDetailsResponse
 	require.NoError(t, serverutils.GetJSONProto(s, "/_admin/v1/databases/test", &resp))
 	assert.Nil(t, resp.Stats, "No Stats unless we ask for them explicitly.")
 
-	nodeIDs := []roachpb.NodeID{1, 2, 3}
+	nodeIDs := tc.NodeIDs()
 	testutils.SucceedsSoon(t, func() error {
 		var resp serverpb.DatabaseDetailsResponse
 		require.NoError(t, serverutils.GetJSONProto(s, "/_admin/v1/databases/test?include_stats=true", &resp))
@@ -68,11 +80,8 @@ func TestAdminAPIDatabaseDetails(t *testing.T) {
 		}
 		assert.Equal(t, nodeIDs, resp.Stats.NodeIDs, "NodeIDs")
 
-		// TODO(todd): Find a way to produce a non-zero value here that doesn't require writing a million rows.
-		// Giving the test nodes on-disk stores (StoreSpec{inMemory: false, Path: "..."}) didn't work and was slow.
-		// Having a number would be nice here for completeness, but the real RangeCount above gives us enough
-		// confidence in the implementation in the meantime.
-		assert.Equal(t, uint64(0), resp.Stats.ApproximateDiskBytes, "ApproximateDiskBytes")
+		// We've flushed data so this estimation should be non-zero.
+		assert.Positive(t, resp.Stats.ApproximateDiskBytes, "ApproximateDiskBytes")
 
 		return nil
 	})

